pipeline/src/pipeline.rs

```diff
@@ -807,132 +511,379 @@ impl<T: FieldElement> Pipeline<T> {
+    }
+
+    fn compute_witness(&mut self) -> Result<Columns<T>, Vec<String>> {
+        self.log("Deducing witness columns...");
+
+        let fixed_cols = self.fixed_cols().unwrap();
+        let pil = self.optimized_pil().unwrap();
+
+        let witness = (pil.constant_count() == fixed_cols.len()).then(|| {
+            let start = Instant::now();
+            let external_witness_values =
+                std::mem::take(&mut self.arguments.external_witness_values);
+            let query_callback = self
+                .arguments
+                .query_callback
+                .take()
+                .unwrap_or_else(|| Arc::new(powdr_executor::witgen::unused_query_callback()));
+            let witness = powdr_executor::witgen::WitnessGenerator::new(
+                &pil,
+                &fixed_cols,
+                query_callback.borrow(),
+            )
+            .with_external_witness_values(external_witness_values)
+            .generate();
+
+            self.log(&format!("Took {}", start.elapsed().as_secs_f32()));
+            witness
+        });
+
+        self.maybe_write_witness(&fixed_cols, &witness)?;
+
+        Ok(witness.unwrap())
+    }
+
+    pub fn witness(&mut self) -> Result<Rc<Columns<T>>, Vec<String>> {
+        if self.artifact.witness.is_none() {
+            self.artifact.witness = Some(Rc::new(self.compute_witness()?));
+        }
+
+        Ok(self.artifact.witness.as_ref().unwrap().clone())
+    }
+
+    fn compute_proof(&mut self) -> Result<Proof, Vec<String>> {
+        let pil = self.optimized_pil().unwrap();
+        let fixed_cols = self.fixed_cols().unwrap();
+        let witness = self.witness().unwrap();
+
+        let backend = self
+            .arguments
+            .backend
+            .expect("backend must be set before calling proving!");
+        let factory = backend.factory::<T>();
+
+        // Opens the setup file, if set.
+        let mut setup = self
+            .arguments
+            .setup_file
+            .as_ref()
+            .map(|path| BufReader::new(fs::File::open(path).unwrap()));
+
+        // Opens the verification key file, if set.
+        let mut vkey = self
+            .arguments
+            .vkey_file
+            .as_ref()
+            .map(|path| BufReader::new(fs::File::open(path).unwrap()));
+
+        /* Create the backend */
+        let backend = factory
+            .create(
+                pil.borrow(),
+                &fixed_cols[..],
+                self.output_dir(),
+                setup.as_io_read(),
+                vkey.as_io_read(),
+            )
+            .unwrap();
+
+        // Reads the existing proof file, if set.
+        let existing_proof = self
+            .arguments
+            .existing_proof_file
+            .as_ref()
+            .map(|path| fs::read(path).unwrap());
+
+        // Even if we don't have all constants and witnesses, some backends will
+        // still output the constraint serialization.
```

Comment:
this would already have failed above because we request both fixed_cols and witness. I'll just remove the comment

Reply:
I mean, we could still support it, because requesting the fixed cols should be fine and the witness could only be requested if the fixed columns are complete. But if there is no use-case, even better if the code is simpler :)